base/sort.jl

```diff
@@ -652,13 +669,203 @@ function sort!(v::AbstractVector, lo::Integer, hi::Integer, a::PartialQuickSort,
+        # if CHUNK_SIZE = 8, counts[2] is bucket 0x00 and counts[257] is bucket 0xff.
+        counts .= 0
+        for k in lo:hi
+            x = v[k]                  # lookup the element
+            i = (x >> shift)&MASK + 2 # compute its bucket's index for this pass
+            counts[i] += 1            # increment that bucket's count
+        end
+
+        counts[1] = lo                # set target index for the first bucket
+        cumsum!(counts, counts)       # set target indices for subsequent buckets
+        # counts[1:MASK+1] now stores indices where the first member of each bucket
+        # belongs, not the number of elements in each bucket. We will put the first element
+        # of bucket 0x00 in t[counts[1]], the next element of bucket 0x00 in t[counts[1]+1],
+        # and the last element of bucket 0x00 in t[counts[2]-1].
+
+        for k in lo:hi # Is this iteration slower than it could be?
+            x = v[k]                  # lookup the element
+            i = (x >> shift)&MASK + 1 # compute its bucket's index for this pass
+            j = counts[i]             # lookup the target index
+            t[j] = x                  # put the element where it belongs
+            counts[i] = j + 1         # increment the target index for the next
+        end                           #  ↳ element in this bucket
+
+        v, t = t, v # swap the now sorted destination list t back into the primary list v
+
+    end
+
+    v
+end
+
+function sort!(v::AbstractVector{<:Bool}, lo::Integer, hi::Integer, a::AdaptiveSort, o::Ordering)
+    first = lt(o, false, true) ? false : lt(o, true, false) ? true : return v
+    count = 0
+    for i in lo:hi
+        if v[i] == first
+            count += 1
+        end
+    end
+    v[lo:lo+count-1] .= first
+    v[lo+count:hi] .= !first
+    v
+end
+
+maybe_unsigned(x::Integer) = x # this is necessary to avoid calling unsigned on BigInt
+maybe_unsigned(x::Union{Int8, Int16, Int32, Int64, Int128}) = unsigned(x)
+function _extrema(v::AbstractArray, lo::Integer, hi::Integer, o::Ordering)
+    mn = mx = v[lo]
+    while lo < hi
+        lo += 1
+        vi = v[lo]
+        lt(o, vi, mn) && (mn = vi)
+        lt(o, mx, vi) && (mx = vi)
+    end
+    mn, mx
+end
+function sort!(v::AbstractVector, lo::Integer, hi::Integer, a::AdaptiveSort, o::Ordering)
+    # if the sorting task is unserializable, then we can't radix sort or sort_int_range!
+    # so we skip straight to the fallback algorithm which is comparison based.
+    U = Serial.Serializable(eltype(v), o)
+    U === nothing && return sort!(v, lo, hi, a.fallback, o)
+
+    # to avoid introducing excessive detection costs for the trivial sorting problem
+    # and to avoid overflow, we check for small inputs before any other runtime checks
+    hi <= lo && return v
+    ln = maybe_unsigned(hi-lo) # ln == length(v[lo:hi])-1. Lower number = fewer letters :)
```

Comment:
rename `ln` to `len`

Reply:
I'll do that that. Is it okay that `len` is a proxy for `length-1`?

Reply:
`lenm1` maybe?

Reply:
Ugly, but accurate. The concept is a bit ugly. I think `lenm1` is probably best (`lengthm1` is misleading because that implies `length(v)`). The reason I'm using this at all is computing `hi-lo+1` doesn't substantively change dispatch, has a (detectable for very small arrays?) cost, and, most importantly, risks overflow.

Reply:
Following this pattern would entail changing `rln` to `rangelenm1` and `urln` to `urangelenm1` or `unsigned_range_len_m1`. I want to be consistent, descriptive, and ideally not use 4 words.

Reply:
from statistics, `range=max-min`, so `range_len_m1` can just be `range`. The reason `sort_int_range!` uses `rangelen` is because it is allocating an array of length `rangelen`, and `rangelen` is a proxy for `range + 1`.

In conclusion
`ln` -> `lenm1`
`rln` -> `range`
and better still:
`mn -> v_min`
`mx -> v_max`
`rln -> v_range`
`umn -> u_min`
`umx -> u_max`
`urln -> u_range`

Reply:
"range" should rather be called "span", right?

Reply:
yeah

Reply:
The concept I'm referring to with the variable name `range` is `max-min`. I believe that this is the unambiguous statistical definition of range, while span is a more ambiguous term. For this reason, I chose to follow @JeffBezanson's choice in 2016 to use the term range for this concept. (Jeff was referring to `max-min+1` and called it `rangelen`.)

Is there some commonly accepted definition of span or range that I'm not aware of? Why do y'all prefer span?

https://en.wikipedia.org/wiki/Range
https://en.wikipedia.org/wiki/Range_(statistics)
https://en.wikipedia.org/wiki/Span

Reply:
Nevermind, I thought this was a standard term but looks like it isn't. We have a function `span` in StatsBase, but it returns... a `min:max` range. :-D